### ISSUE TITLE
OS X: Unable to compile macro with conda ROOT+python

### ISSUE THREAD

**Issue by kreczko** [NLeSC/root-conda-recipes#7]
Hi,

Conda env: https://github.com/kreczko/ntuple/blob/master/.travis.yml
OS: OS X (10.10)
ROOT: 6
Python: 2.7

I am trying to compile a [ROOT macro (example.C)](https://gist.github.com/kreczko/35b022c7f711e8981fc5#file-root-macro) within python to get access to `std::vector` (to be exact rootpy tries this automatically when calling `stl.vector`).
The command for compilation is `python -c "import ROOT;print ROOT.gSystem.CompileMacro('example.C', 'kv-')"`
and it fails with

```
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file '/software/miniconda/envs/ntuple/lib/python2.7/lib-dynload/_locale.so' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in <ACLiC>: Compilation failed!
```

The full command with the full list of libraries can be seen [here](https://gist.github.com/kreczko/35b022c7f711e8981fc5#file-not-working). As you can see this includes libraries in `/software/miniconda/envs/ntuple/lib/python2.7/lib-dynload/`. If I remove them by hand from the linking procedure, the macro compiles.

The macro also compiles with a ROOT version compiled against the system python (nothing from conda). This version has been configured with

```
   ./configure
     "--enable-roofit"
     "--enable-minuit2"
     "--enable-builtin-freetype"
     "--with-clang"
arch=darwin
chip=x86_64
rele=14.0.0
macosx64
```

My assumption is that something is wrong as to how ROOT is compiled against the conda python, but I am not sure.

**Comment by kreczko** [comment]
Just tried this on my Ubuntu 14.04 machine and it works well here (same env recipe).
Something wrong with the OS X recipe (updated title)

**Comment by remenska** [comment]
Hi @kreczko, thanks for reporting. Indeed, I'm aware the OSX binary is flacky. I have very little experience myself with OS X libraries/dependencies, is why I keep pushing it to a lower priority. 

I'll come back to this and also your previous reported issue soon (noted). Right now I'm preparing for a PhD defense, so ROOT/conda issues are a bit delayed :)

**Comment by kreczko** [comment]
@remenska Good luck with your defense (https://xkcd.com/1403/)!
In the meantime I will try to understand the conda build system a bit.

**Comment by kreczko** [comment]
Probably not related:
I noticed that the ROOT compilation commands include folders that are not present on the machines.
On OS X: `/Users/vagrant/anaconda/conda-bld/work/workdir/*` 
On Ubuntu: `/media/NewPartition/anaconda3/conda-bld/work/workdir`

Possibly related:
On OS X `clang++` is used for compilation
```shell
which clang++
/usr/bin/clang++
```
On Ubuntu `c++` is used:
```shell
which c++
/software/miniconda/envs/ntuple/bin/c++
```

Under OS X, I guess because python and ROOT are compiled against the conda gcc, it does not work.

Investigating further I see
```shell
# OS X
python -c 'import ROOT; print ROOT.gSystem.GetBuildCompiler()'
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++
# on Ubuntu
python -c 'import ROOT; print ROOT.gSystem.GetBuildCompiler()'
/opt/rh/devtoolset-2/root/usr/bin/c++ # <- does not exist!
```
The return value of `ROOT.gSystem.GetBuildCompiler()` is read from a constant, probably set at build-time. Since it does not exist on Ubuntu it falls back to the conda installed `c++` which is fine for the python libraries.

So I think the solution is either:
 - compile python on OS X against clang++
 - OR compile ROOT on OS X against the gcc install, not clang++

I would prefer the second option in both cases just to make sure ROOT is compiled against the compiler that is included in the dependencies (https://github.com/remenska/root-conda-recipes/blob/master/root6/meta.yaml#L50) to prevent future hick-ups.

**Comment by remenska** [comment]
Thanks @kreczko for the investigation. I remember trying (unsuccessfully) to build ROOT with gcc. 
Indeed the right way to do it is ship gcc along with ROOT compiled with the same gcc version, just like with the Linux binary. There are issues to be resolved to make the binary portable (glibc, the gui/graphics etc) on OSX. 

Meanwhile, I thought the strong requirement of "no sudo access" on OSX can be lifted, and we can assume that most people with a Macebook actually own it, and have xcode installed. Hence the system clang++.
This would alleviate the need to ship many standard dependency libraries which are already available within the framework. 

I appreciate any help when I come back to the issue more closely.

Thanks again!

**Comment by remenska** [comment]
Btw, @kreczko have you tried conda ROOT5 on MacOS?

**Comment by arogozhnikov** [comment]
Hi, Daniela. First, thanks for this super recipes.
I'm trying now to get it working on OS X. 

Travis tests pass only if I delete packages like libpng etc.,otherwise strange errors appear on mac os x.
(btw, note that `--force` is required now during removing packages - otherwise conda deletes everything, including ROOT)

At the same time, I can't remove, say, libpng, because otherwise matplotlib doesn't work.

Here you can find a minimal testing example:
https://github.com/arogozhnikov/root-conda-recipes/blob/master/.travis.yml

**Comment by kreczko** [comment]
Hi Daniela,

Finally got around to it. I created a new environment with and installed ROOT5.
And indeed, ROOT5 works!

**Comment by remenska** [comment]
@arogozhnikov Thanks a lot for the `--force` hint, conda has changed this in the meantime it seems. I have updated the travis builds. 

@kreczko compiling OSX ROOT6 with GCC is something I tried a long time ago, but compilation failed for various reasons, and making it portable is more effort (peeling off more dependencies than you would get under a linux distro, this is why the OSX ROOT recipe doesn't have GCC as a dependency..it was compiled with clang++ in fact).
I will give it a nudge again.
